fuel-vm/src/interpreter/metadata.rs

```diff
@@ -108,53 +108,41 @@ pub(crate) fn metadata(
     chain_id: ChainId,
     tx_offset: Word,
 ) -> SimpleResult<()> {
-    let external = context.is_external();
     let args = GMArgs::try_from(imm)?;
+    let parent = context
+        .is_internal()
+        .then(|| frames.last().map(|f| f.registers()[RegId::FP]))
+        .flatten();
 
-    if external {
-        match args {
-            GMArgs::GetVerifyingPredicate => {
-                *result = context
-                    .predicate()
-                    .map(|p| p.idx() as Word)
-                    .ok_or(PanicReason::TransactionValidity)?;
-            }
-
-            GMArgs::GetChainId => {
-                *result = chain_id.into();
-            }
-
-            GMArgs::TxStart => {
-                *result = tx_offset;
-            }
-
-            _ => return Err(PanicReason::ExpectedInternalContext.into()),
+    match (args, parent) {
+        (GMArgs::GetVerifyingPredicate, None) => {
+            *result = context
+                .predicate()
+                .map(|p| p.idx() as Word)
+                .ok_or(PanicReason::TransactionValidity)?;
         }
-    } else {
-        let parent = frames
-            .last()
-            .map(|f| f.registers()[RegId::FP])
-            .expect("External context will always have a frame");
 
-        match args {
-            GMArgs::IsCallerExternal => {
-                *result = (parent == 0) as Word;
-            }
+        (GMArgs::GetChainId, _) => {
+            *result = chain_id.into();
+        }
 
-            GMArgs::GetCaller if parent != 0 => {
-                *result = parent;
-            }
+        (GMArgs::BsaeAssetId, _) => {
+            *result = tx_offset;
+        }
 
-            GMArgs::GetChainId => {
-                *result = chain_id.into();
-            }
+        (GMArgs::TxStart, _) => {
+            *result = tx_offset;
+        }
 
-            GMArgs::TxStart => {
-                *result = tx_offset;
-            }
+        (GMArgs::GetCaller, Some(parent)) if parent != 0 => {
+            *result = parent;
+        }
 
-            _ => return Err(PanicReason::ExpectedInternalContext.into()),
+        (GMArgs::IsCallerExternal, Some(parent)) => {
+            *result = (parent == 0) as Word;
         }
+
+        _ => return Err(PanicReason::ExpectedInternalContext.into()),
```

Comment:
This doesn't seem very future-proof, as the context error as a catch-all could happen even when it's an unrelated error.

Reply:
Even before, it could actually occur during both contexts, and definitely doesn't seem to be the correct error. I'll rework this part.

Reply:
Fixed in 3fe06ca. I changed the match to an exhaustive one and added a new panic variant `ExpectedNestedCaller` for the relevant case.